store/pg_engine.go

```diff
@@ -403,47 +419,137 @@ func getCutoffVersion(releaseMode common.ReleaseMode) (semver.Version, error) {
 		return semver.Version{}, fmt.Errorf("Migration path %s has no minor version", minorPathPrefix)
 	}
 	minorVersion := versions[len(versions)-1]
-	upperMinorVersion := minorVersion
-	upperMinorVersion.Minor = upperMinorVersion.Minor + 1
 
-	patchPathPrefix := fmt.Sprintf("migration/%s/%d.%d", releaseMode, minorVersion.Major, minorVersion.Minor)
+	patchPathPrefix := fmt.Sprintf("migration/%s/%d.%d", common.ReleaseModeRelease, minorVersion.Major, minorVersion.Minor)
 	names, err = fs.Glob(migrationFS, fmt.Sprintf("%s/*.sql", patchPathPrefix))
 	if err != nil {
 		return semver.Version{}, err
 	}
-	patchVersions, err := getPatchVersions(minorVersion, upperMinorVersion, semver.Version{} /* currentVersion */, names)
+	patchVersions, err := getPatchVersions(minorVersion, semver.Version{} /* currentVersion */, names)
 	if err != nil {
 		return semver.Version{}, err
 	}
-	// This should only happen in the dev environment
 	if len(patchVersions) == 0 {
-		if releaseMode == common.ReleaseModeDev {
-			return minorVersion, nil
-		}
 		return semver.Version{}, fmt.Errorf("Migration path %s has no patch version", patchPathPrefix)
 	}
 	return patchVersions[len(patchVersions)-1].version, nil
 }
 
+func migrateDev(ctx context.Context, d dbdriver.Driver, serverVersion, databaseName string, cutoffSchemaVersion semver.Version, histories []*dbdriver.MigrationHistory, l *zap.Logger) error {
+	devMigrations, err := getDevMigrations()
+	if err != nil {
+		return err
+	}
+
+	var migrations []devMigration
+	// Skip migrations that are already applied, otherwise the migration reattempt will most likely to fail with already exists error.
+	for _, m := range devMigrations {
+		if migrationExists(m.statement, histories) {
+			l.Info(fmt.Sprintf("Skip migrating dev migration file %s that's already migrated.", m.filename))
+		} else {
+			migrations = append(migrations, m)
+		}
+	}
+	if len(migrations) == 0 {
+		l.Info("Skip dev mode migration; no new version.")
+		return nil
+	}
+
+	for _, m := range migrations {
+		l.Info(fmt.Sprintf("Migrating dev %s...", m.filename))
+		// We don't use semantic versioning for dev migrations because they are not versioned yet.
+		if _, _, err := d.ExecuteMigration(
+			ctx,
+			&dbdriver.MigrationInfo{
+				ReleaseVersion:        serverVersion,
+				UseSemanticVersion:    true,
+				Version:               cutoffSchemaVersion.String(),
+				SemanticVersionSuffix: fmt.Sprintf("dev-%s", m.version),
```

Comment:
I inclined to use dev%s instead of dev-%s because - is also the separator between semantic version and the suffix

Reply:
Done.